subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyConstraintsIntegrationTest.groovy

```diff
@@ -0,0 +1,247 @@
+
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.ExperimentalFeaturesFixture
+
+/**
+ * This is a variation of {@link OptionalDependenciesIntegrationTest} that tests dependency constraints
+ * declared in the build script (instead of published optional dependencies)
+ */
+class DependencyConstraintsIntegrationTest extends AbstractIntegrationSpec {
+
+    def setup() {
+        ExperimentalFeaturesFixture.enable(settingsFile)
+        settingsFile << "rootProject.name = 'test'"
+    }
+
+    void "dependency constraint is ignored when feature not enabled"() {
+        given:
+        mavenRepo.module("org", "foo", '1.0').publish()
+        mavenRepo.module("org", "foo", '1.1').publish()
+
+        // Do not enable feature
+        settingsFile.text = """
+            rootProject.name = 'test'
+"""
+        buildFile << """
+            repositories {
+                maven { url "${mavenRepo.uri}" }
+            }
+            configurations {
+                conf
+            }
+            dependencies {
+                conf 'org:foo:1.0'
+                constraints {
```

Comment:
:o: Since experimental features are disabled, I wonder if it would be interesting to issue a warning here, saying "hey, you're using the `constraints` DSL, but you haven't enabled experimental features using ..., this is going to be ignored"

Reply:
I think that's a good point for any feature that creates _pending_ nodes (formally known as _optional_). I think we should address this later, but soon, as part of  https://github.com/gradle/gradle-private/issues/1030